[lldb] [Mach-O] ProcessMachCore needs to strip TBI data from addrs

jasonmolenda opened the pull request:
Darwin AArch64 application processors are run with Top Byte Ignore mode enabled so metadata may be stored in the top byte, it needs to be ignored when reading/writing memory.  David Spickett handled this already in the base class Process::ReadMemory but ProcessMachCore overrides that method (to avoid the memory cache) and did not pick up the same change.  I add a test case that creates a pointer with metadata in the top byte and dereferences it with a live process and with a corefile.

rdar://123784501

jasonmolenda committed:
[lldb] [Mach-O] ProcessMachCore needs to strip TBI data from addrs

Darwin AArch64 application processors are run with Top Byte Ignore
mode enabled so metadata may be stored in the top byte, it needs
to be ignored when reading/writing memory.  David Spickett handled
this already in the base class Process::ReadMemory but ProcessMachCore
overrides that method (to avoid the memory cache) and did not pick
up the same change.  I add a test case that creates a pointer with
metadata in the top byte and dereferences it with a live process and
with a corefile.

rdar://123784501

DavidSpickett commented on lldb/source/Plugins/Process/mach-core/ProcessMachCore.cpp:
```diff
@@ -652,7 +652,7 @@ size_t ProcessMachCore::ReadMemory(addr_t addr, void *buf, size_t size,
   // Don't allow the caching that lldb_private::Process::ReadMemory does since
   // in core files we have it all cached our our core file anyway.
-  return DoReadMemory(addr, buf, size, error);
+  return DoReadMemory(FixAnyAddress(addr), buf, size, error);
```
I'm tempted to move this up into `lldb/include/lldb/Target/PostMortemProcess.h` but not all of them fix addresses yet, and it's better changes like that are done explicitly by people who use those classes. If we get to a point where everyone is fixing the address then we can pull this up.

DavidSpickett commented on lldb/test/API/macosx/tbi-honored/TestTBIHonored.py:
```diff
@@ -0,0 +1,49 @@
+"""Test that lldb on Darwin ignores metadata in the top byte of addresses."""
+
+import os
```
These seem to be unused.

DavidSpickett commented on lldb/test/API/macosx/tbi-honored/TestTBIHonored.py:
```diff
@@ -0,0 +1,49 @@
+class TestTBIHonored(TestBase):
+    @no_debug_info_test
+    @skipUnlessDarwin
+    @skipIf(archs=no_match(["arm64", "arm64e"]))
```
This isn't a live process, do we need to be on Darwin arm64?

DavidSpickett commented on lldb/test/API/macosx/tbi-honored/TestTBIHonored.py:
```diff
@@ -0,0 +1,49 @@
+        self.expect("frame variable *pb.p", substrs=["15"])
+        self.expect("expr -- *(int*)0x%x" % addr, substrs=["15"])
+
+    def test(self):
```
I think the test framework is going to run this method as the test case, so put the decorators on this instead.

DavidSpickett commented on lldb/test/API/macosx/tbi-honored/TestTBIHonored.py:
```diff
@@ -0,0 +1,49 @@
+"""Test that lldb on Darwin ignores metadata in the top byte of addresses."""
```
Oh I get it, you're running a process first then making a core file and expecting to get the same output.

This needs to be made clear at the top of the file at least, with some extra comments later too.

DavidSpickett commented on lldb/test/API/macosx/tbi-honored/TestTBIHonored.py:
```diff
@@ -0,0 +1,49 @@
+    @skipUnlessDarwin
+    @skipIf(archs=no_match(["arm64", "arm64e"]))
+    @skipIfRemote
+    def do_variable_access_tests(self, frame):
```
Just in case we mess up really badly, it is worth checking that pb.bytes[7] is in fact 0xfe first.

That would also functions as documentation implying "despite this being 0xfe we can... <rest of expects>".

DavidSpickett commented on lldb/test/API/macosx/tbi-honored/main.c:
```diff
@@ -0,0 +1,13 @@
+union ptrbytes {
+  int *p;
+  uint8_t bytes[8];
+};
```
I like this way of avoiding shifts.

jasonmolenda committed:
Address David Spickett's feedback on the testcase

Lots of fixes to the test case for correctness and
to make it clearer what it is testing.

jasonmolenda commented:
Thanks for all the helpful comments as always, David.  I updated the test case to fix the issues you identified and added comments that I think make it clearer that it is testing both a live process and a corefile.

DavidSpickett reviewed (approved):
LGTM
